subprojects/tooling-api/src/testFixtures/groovy/org/gradle/integtests/tooling/fixture/ToolingApiDistributionResolver.groovy

```diff
@@ -24,30 +24,43 @@ import org.gradle.integtests.fixtures.RepoScriptBlockUtil
 import org.gradle.integtests.fixtures.executer.IntegrationTestBuildContext
 import org.gradle.integtests.fixtures.executer.CommitDistribution
 import org.gradle.testfixtures.ProjectBuilder
+import org.gradle.testfixtures.internal.ProjectBuilderImpl
+
+import javax.annotation.Nonnull
+
+/**
+ * Note that the resolver instance must be closed after use in order to release resources.
+ */
+class ToolingApiDistributionResolver implements AutoCloseable {
+
+    private ProjectInternal project = null
```

Comment:
I suppose you do the lazy initialization here so we don't need to clean up the resolvers which never resolve anything? Or is there another reason?

I think it would be good if we could keep the fields final if there aren't any huge downsides.

Reply:
Yes, I explicitly do it lazily. Firstly, as you mentioned, to avoid allocation of all services for the ProjectBuilder project if they are not needed. Secondly, to ensure that the Resolver itself can be safely created without creating the project immediately, so that if any exception is thrown, the project would get disposed of. If we create the whole project in the Resolver's constructor we risk the leakage again, because the constructor is usually called outside of the try block (when try-with-resources is desugared).

Why do we need/really want to have these fields final in a text fixture? When I skimmed through the usages, they seemed safe from the concurrency perspective.

Reply:
I think the class is used only in two places and then closed directly afterwards. How about keeping those fields final, making the constructor private and have a static method on the class which takes a callable and allows you to do the things you need to do?

Reply:
That sounds good. Also similar to how we decided to handle a similar thing in Kotlin

Reply:
Done